### PR TITLE
Globe view further optimizations

### PR DESCRIPTION
**Pull request by akoylasar**
This PR introduces some more optimizations for the globe view. 
The change here specifically implements a shared vertex buffer which is initialized once at map creation and then reused across all tiles when globe view is enabled. This is the first approach that's been described in the issue [here](https://github.com/mapbox/mapbox-gl-js/issues/11557#issuecomment-1058047339).

### Details:

Previously  the same vertex shader was used to render all tiles. This change necessitates the introduction of a separate vertex shader (via a *shader define*) which is slightly more expensive since the computations that were previously done on CPU per tile on each zoom level change, are now computed per frame on GPU. The code has been updated accordingly to minimize the number of program switches during rendering. The performance gains overall justify this additional complexity.

The screenshot below also highlights a manually profiled session where tile buffer creation and GPU upload  was happening on zoom changes in previous implementation (left) vs this implementation where the shared buffer is created and uploaded only once. 

num_calls

Update: Initially this PR also introduced a mechanism to reduce the vertex count of the tile patches. This resulted in some slight visual regressions at certain zoom levels. The [decision](https://github.com/mapbox/mapbox-gl-js/pull/11560#issuecomment-1066936050) was to rather introduce this in a separate PR.

cc @mpulkki-mapbox @karimnaaji  

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] Add screenshot and show significantly reduced number of calls to upload buffers.
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog> Use a single shared buffer across all globe tiles to boost rendering performance and avoid jank. </changelog>`

### CODE REVIEW

**Comment by akoylasar** [comment]
> Have you considered also adjusting the LOD based on the latitude? Going level down as tile-ids go towards the pole, so that the LOD you introduced is always matching at the equator and the resolution get reduced as we move to the pole.

@karimnaaji  Yes indeed I was considering this! The function `getTileLod` in `globe_util.js` can be modified to take latitude into account as well. I haven't played around with this yet but it's a good next step.

File: src/geo/projection/globe_util.js
Comment (akoylasar) [reply src/geo/projection/globe_util.js:477]: @mpulkki-mapbox I was aware of the issue of tiles not being watertight but this problem could still manifest itself before this change and with POT values although with POT it would be harder to notice it. My goal was to find a good set of values for the LOD table to mitigate this issue. Adding tile latitude to the equation would make this even more challenging cc @karimnaaji. I will investigate this further.
I will also investigate the issue with the wireframe thanks for the catch!

Comment (akoylasar) [reply src/geo/projection/globe_util.js:477]: @mpulkki-mapbox regarding the code snippet currently the LOD table is chosen to have the same size as the minimum globe zoom threshold. This is enforced via an assertion so that the ranges are aligned.

Comment (mpulkki-mapbox) [reply src/geo/projection/globe_util.js:477]: @akoylasar, good point, the issue with borders not being watertight is not a regression :)

Comment (akoylasar) [reply src/geo/projection/globe_util.js:477]: @mpulkki-mapbox unfortunately though the wireframe issue is caused by precision loss. This in fact can become seriously noticeable when terrain is enabled as seen below:
terrain_globeview_precisionloss
I will try to think if there's a way to preserve precision because this in fact could be a deal breaker.

Comment (mpulkki-mapbox) [reply src/geo/projection/globe_util.js:477]: I don't think this is a precision issue. Looks like that the elevation sampling is not setup correctly after the latest changes

Comment (akoylasar) [reply src/geo/projection/globe_util.js:477]: @mpulkki-mapbox I haven't modified the elevation code when it comes to rendering the non-pole tiles. Perhaps I have messed up something and have overlooked it. Could you please point me to where you think the issue is?

Comment (akoylasar) [reply src/geo/projection/globe_util.js:477]: Ha! I found the issue and you were right it wasn't due to precision issues so I think I jumped to conclusion somewhat quickly. Anyways the issue was this:
```
vec4 globe = u_globe_matrix * vec4(globe_pos + up_vector.xyz * height, 1.0);
```
previously `globe_pos` was of course normalized but this is no longer the case since we calculate it in the shader. I simply overlooked the fact that the `up_vector` is also in the same space. What should have been done however is the following:
```
vec4 vv = u_normalized_globe_matrix * vec4(up_vector.xyz * height, 0.0);
vec4 globe = u_globe_matrix * vec4(globe_pos, 1.0);
globe = globe + vv;
```
It took me lots of debugging to figure notice this.. one of those cases! Sorry for the confusion and thanks again!

Comment (mpulkki-mapbox) [reply src/geo/projection/globe_util.js:477]: Good catch! The shader uniform `up_vector` contains both a direction and a scale for converting elevation values (in meters) to whatever the tile coordinate space is. I think you could even simplify this a bit by just changing the value of `u_tile_up_scale` after invoking `terrain.setupElevationDraw` to a correct conversion value (meters-to-pixels).

**Comment by akoylasar** [comment]
@karimnaaji I update the `getTileLod` function to take latitude into account. There is one important thing to note in that for this to actually work we need to have an LOD table that has different entries at each zoom level. This is because the shared buffer is created once and we basically choose which resolution to pick only. Currently the table is such the latitude will affect the resolution at zoom level 3-4. Of course we're free to choose a different table but we need to agree what looks best. Moreover some of the test image references need to be updated accordingly. Currently all tests pass and everything seems to look fine. The tears are also far less noticeable. The image below showcases the updated implmentation:
latitude_sensitivity
If you look closely the image on the left has higher number of subdivisions that the one on the right (from current implementation which takes latitude into account).
cc @mpulkki-mapbox

**Comment by akoylasar** [comment]
At this point if you have no further concerns or think that the LOD table is fine I would suggest that I proceed with providing a report from a custom benchmark to compare this implementation with the previous one. I am open for suggestions regarding the said benchmark. 
cc @karimnaaji @mpulkki-mapbox

**Comment by mpulkki-mapbox** [comment]
> @karimnaaji I updated the `getTileLod` function to take latitude into account. There is one important thing to note in that for this to actually work we need to have an LOD table that has different entries at each zoom level. This is because the shared buffer is created once and we basically choose which resolution to pick only. Currently the table is such that the latitude will affect the resolution at zoom level 3-4. Of course we're free to choose a different table but we need to agree what looks best. Moreover some of the test image references need to be updated accordingly. Currently all tests pass and everything seems to look fine. The tears are also far less noticeable. The image below showcases the updated implmentation: latitude_sensitivity If you look closely the image on the left has higher number of subdivisions that the one on the right (from current implementation which takes latitude into account). Note however that this image was taken with the original LOD table. cc @mpulkki-mapbox

The right image seems to have more horizontal subdivisions than the left one (ratio of cells being 9:8) or did I understood something wrong? It is also unclear which of the two is from this PR.

File: src/terrain/terrain.js
Comment (mpulkki-mapbox) [inline src/terrain/terrain.js:592]: Should we use some more descriptive name for the parameter?

File: src/geo/projection/globe_util.js
Comment (mpulkki-mapbox) [inline src/geo/projection/globe_util.js:47]: What kind of camera parameters do I need to use to see the LOD in action?

**Comment by mpulkki-mapbox** [comment]
Incrementing `GLOBE_ZOOM_THRESHOLD_MIN` and `GLOBE_ZOOM_THRESHOLD_MAX` triggers an assertion that seems to be a regression

File: src/terrain/terrain.js
Comment (akoylasar) [reply src/terrain/terrain.js:592]: sure. How about `useDenormalizedUpVector`?

File: src/geo/projection/globe_util.js
Comment (akoylasar) [reply src/geo/projection/globe_util.js:47]: @mpulkki-mapbox I fixed a bug that I had accidentally introduced when calculating the tile center latitude used in the LOD calculation. The top image is what came before taking latitude into account. The lower image takes latitude into account:
without_latitude_change
with_latitude_change
Notice how tiles closer to the pole have fewer subdivisions. To take things even further we could subdivide tiles rather based on their distance from map center. It's perhaps a good thing to do this in a separate PR in future. cc @karimnaaji

**Comment by akoylasar** [comment]
> > @karimnaaji I updated the `getTileLod` function to take latitude into account. There is one important thing to note in that for this to actually work we need to have an LOD table that has different entries at each zoom level. This is because the shared buffer is created once and we basically choose which resolution to pick only. Currently the table is such that the latitude will affect the resolution at zoom level 3-4. Of course we're free to choose a different table but we need to agree what looks best. Moreover some of the test image references need to be updated accordingly. Currently all tests pass and everything seems to look fine. The tears are also far less noticeable. The image below showcases the updated implmentation: latitude_sensitivity If you look closely the image on the left has higher number of subdivisions that the one on the right (from current implementation which takes latitude into account). Note however that this image was taken with the original LOD table. cc @mpulkki-mapbox
> 
> The right image seems to have more horizontal subdivisions than the left one (ratio of cells being 9:8) or did I understood something wrong? It is also unclear which of the two is from this PR.



> Incrementing `GLOBE_ZOOM_THRESHOLD_MIN` and `GLOBE_ZOOM_THRESHOLD_MAX` triggers an assertion that seems to be a regression

@mpulkki-mapbox this is actually intentional. I put this assertion to enforce the alignment of LOD table size with the `GLOBE_ZOOM_THRESHOLD_MIN`. When changing `GLOBE_ZOOM_THRESHOLD_MIN` we should decrease/increase the number of entries in `GLOBE_VERTEX_GRID_SIZE_LOD_TABLE` accordingly. This keeps the implementation simple and also gives us the opportunity to decide what LOD to associate with the newly added zoom level should we decide to have a higher threshold for instance.
In short
```
// current thresholds:
const GLOBE_ZOOM_THRESHOLD_MIN = 5;
const GLOBE_ZOOM_THRESHOLD_MAX = 6;
// current table
const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32]; // size == 5
```

if we choose to increment the thresholds this is how things should look like:
```
const GLOBE_ZOOM_THRESHOLD_MIN = 7;
const GLOBE_ZOOM_THRESHOLD_MAX = 8;
const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32, 32, 32]; // size  == 7
```

**Comment by mpulkki-mapbox** [comment]
> > Incrementing `GLOBE_ZOOM_THRESHOLD_MIN` and `GLOBE_ZOOM_THRESHOLD_MAX` triggers an assertion that seems to be a regression
> 
> @mpulkki-mapbox this is actually intentional. I put this assertion to enforce the alignment of LOD table size with the `GLOBE_ZOOM_THRESHOLD_MIN`. When changing `GLOBE_ZOOM_THRESHOLD_MIN` we should decrease/increase the number of entries in `GLOBE_VERTEX_GRID_SIZE_LOD_TABLE` accordingly. This keeps the implementation simple and also gives us the opportunity to decide what LOD to associate with the newly added zoom level should we decide to have a higher threshold for instance. In short
> 
> ```
> // current thresholds:
> const GLOBE_ZOOM_THRESHOLD_MIN = 5;
> const GLOBE_ZOOM_THRESHOLD_MAX = 6;
> // current table
> const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32]; // size == 5
> ```
> 
> if we choose to increment the thresholds this is how things should look like:
> 
> ```
> const GLOBE_ZOOM_THRESHOLD_MIN = 7;
> const GLOBE_ZOOM_THRESHOLD_MAX = 8;
> const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32, 16, 16]; // size  == 7
> ```
> 
> cc @SnailBones

would it be a better idea to choose one of the values from the array in this case?

**Comment by mpulkki-mapbox** [comment]
Looks good, but there's a one thing that should be addressed before merging. Flickering is very noticeable now especially with darker styles due to grid meshes of different densities used next to each other. Previously LOD was determined as a function of the distance between center point and the tile location meaning that there were no or less flickering at the center of the screen.

https://user-images.githubusercontent.com/55925868/157926375-c9e3ba2b-8269-4eb7-862e-2f9e47ac6fb8.mp4

**Comment by akoylasar** [comment]
> Looks good, but there's a one thing that should be addressed before merging. Flickering is very noticeable now especially with darker styles due to grid meshes of different densities used next to each other. Previously LOD was determined as a function of the distance between center point and the tile location meaning that there were no or less flickering at the center of the screen.
> 
>  tile_cover_flickering.mp4

Thanks for checking this @mpulkki-mapbox ! Luckily n the  meantime I was playing with a different way of picking LOD based on distance of tile from map center and I think it fixes the tearing significantly. I will test this with a dark style and report back here.

**Comment by akoylasar** [comment]
> > > Incrementing `GLOBE_ZOOM_THRESHOLD_MIN` and `GLOBE_ZOOM_THRESHOLD_MAX` triggers an assertion that seems to be a regression
> > 
> > 
> > @mpulkki-mapbox this is actually intentional. I put this assertion to enforce the alignment of LOD table size with the `GLOBE_ZOOM_THRESHOLD_MIN`. When changing `GLOBE_ZOOM_THRESHOLD_MIN` we should decrease/increase the number of entries in `GLOBE_VERTEX_GRID_SIZE_LOD_TABLE` accordingly. This keeps the implementation simple and also gives us the opportunity to decide what LOD to associate with the newly added zoom level should we decide to have a higher threshold for instance. In short
> > ```
> > // current thresholds:
> > const GLOBE_ZOOM_THRESHOLD_MIN = 5;
> > const GLOBE_ZOOM_THRESHOLD_MAX = 6;
> > // current table
> > const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32]; // size == 5
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > if we choose to increment the thresholds this is how things should look like:
> > ```
> > const GLOBE_ZOOM_THRESHOLD_MIN = 7;
> > const GLOBE_ZOOM_THRESHOLD_MAX = 8;
> > const GLOBE_VERTEX_GRID_SIZE_LOD_TABLE = [64, 64, 64, 64, 32, 16, 16]; // size  == 7
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > cc @SnailBones
> 
> would it be a better idea to choose one of the values from the array in this case?

@mpulkki-mapbox you mean using `32` instead of `16` (which I only put there as an example)? If so yes it should be just fine. If not could you please elaborate a bit more?

**Comment by mpulkki-mapbox** [comment]
> @mpulkki-mapbox you mean using `32` instead of `16` (which I only put there as an example)? If so yes it should be just fine. If not could you please elaborate a bit more?

Could we use some fallback value if the array is not large enough?

**Comment by mpulkki-mapbox** [comment]
> Thanks for checking this @mpulkki-mapbox ! Luckily in the meantime I was playing with a different way of picking LOD based on distance of tile from map center and I think it fixes the tearing significantly. I will test this with a dark style and report back here.
> 
> Update: I checked latest master and I am afraid I see the same behaviour. Peronally I couldn't even notice the difference as they're pretty identical. Did you try to compare them as well?
> 
> An alternative approach is to ensure that all subdivisions from tiles line up. The idea is taking advantage of the fact that on each zoom level the difference in tile zoom level is at most 1. So we may have a tile that has a zoom level of 3 and another one (bit further from camera) with a zoom level of 2. In this case the former tile gets divided say 64 times while the latter one gets 32. This should make the space between tiles watertight. This actually increases vertex count in contrast to what's been done so far.

I do think there's a regression in the visual quality but not in the logic itself. The flickering looks more severe because there's less triangles and hence larger gaps between triangles of adjacent meshes. Few options what to do:
 1. Fix the flickering in this PR.
 2. Split the mesh size reduction into a separate PR and merge the unified mesh code.
 3. Merge the PR as-is and fix the regression in a separate PR.

@karimnaaji what's your opinion on this one?

**Comment by akoylasar** [comment]
> > Thanks for checking this @mpulkki-mapbox ! Luckily in the meantime I was playing with a different way of picking LOD based on distance of tile from map center and I think it fixes the tearing significantly. I will test this with a dark style and report back here.
> > Update: I checked latest master and I am afraid I see the same behaviour. Peronally I couldn't even notice the difference as they're pretty identical. Did you try to compare them as well?
> > An alternative approach is to ensure that all subdivisions from tiles line up. The idea is taking advantage of the fact that on each zoom level the difference in tile zoom level is at most 1. So we may have a tile that has a zoom level of 3 and another one (bit further from camera) with a zoom level of 2. In this case the former tile gets divided say 64 times while the latter one gets 32. This should make the space between tiles watertight. This actually increases vertex count in contrast to what's been done so far.
> 
> I do think there's a regression in the visual quality but not in the logic itself. The flickering looks more severe because there's less triangles and hence larger gaps between triangles of adjacent meshes. Few options what to do:
> 
> 1. Fix the flickering in this PR.
> 2. Split the mesh size reduction into a separate PR and merge the unified mesh code.
> 3. Merge the PR as-is and fix the regression in a separate PR.
> 
> @karimnaaji what's your opinion on this one?

I think option 2 makes more sense. The main can benefit from the shared buffer optimization already so best to merge that part since it's ready. A separate PR also give us an opportunity to discuss the approach to geometry reduction in more detail and ensure that no quality is lost. Hopefully @karimnaaji  also agrees :)

**Comment by karimnaaji** [comment]
>I think option 2 makes more sense. The main can benefit from the shared buffer optimization already so best to merge that part since it's ready. A separate PR also gives us an opportunity to discuss the approach to geometry reduction in more detail and ensure that no quality is lost. Hopefully @karimnaaji also agrees :)

+1 on splitting that PR, the LOD is a great idea and something we should get back to but the benefits of using a global shared mesh are visible and is what hits the CPU workload the most.

>I do think there's a regression in the visual quality but not in the logic itself. The flickering looks more severe because there's less triangles and hence larger gaps between triangles of adjacent meshes.

@akoylasar For future work on the flickering, it might be possible to extend the mesh by an epsilon value so that meshes slightly overlap between tiles. It might be worth a try because the flickering issue seems to be only of a few pixels wide and extending the mesh extent to be slightly bigger might resolve it.

File: src/shaders/globe_raster.vertex.glsl
Comment (karimnaaji) [inline src/shaders/globe_raster.vertex.glsl:5]: nit: Might be a good call to put these in the shader prelude closer to the other `PI` define we have: https://github.com/mapbox/mapbox-gl-js/blob/main/src/shaders/_prelude.glsl#L6

File: src/shaders/globe_raster.vertex.glsl
Comment (karimnaaji) [inline src/shaders/globe_raster.vertex.glsl:58]: Why encoding these in the `mat3` and not in their own uniforms? It should be possible to consolidate these as their own `vec3`. Otherwise, I find it a bit confusing as part of the `mat3` as they aren't affecting the transform at https://github.com/mapbox/mapbox-gl-js/pull/11560/files#diff-184e79e72d9cb77dd6f755ab3fc6f022fde3fee19cd63971da8672ef0bd2a834R60.

**Comment by akoylasar** [comment]
> @akoylasar For future work on the flickering, it might be possible to extend the mesh by an epsilon value so that meshes slightly overlap between tiles. It might be worth a try because the flickering issue seems to be only of a few pixels wide and extending the mesh extent to be slightly bigger might resolve it.

@karimnaaji that's actually a very good idea! I will open a PR for this work soon after I merge this one and try to see if this could alleviate the flickering issue.

**Review by karimnaaji — COMMENTED** [review]
Works great, thanks for looking at this. For the split of the work with the LOD feel free to open a draft PR with the considerations and what needs to happen to move it forward.

I have one nit and a question about splitting the uniforms out of the `mat3` `u_grid_matrix`.

File: src/shaders/globe_raster.vertex.glsl
Comment (akoylasar) [reply src/shaders/globe_raster.vertex.glsl:5]: Yes I think that's a good idea too :)

File: src/shaders/globe_raster.vertex.glsl
Comment (akoylasar) [reply src/shaders/globe_raster.vertex.glsl:58]: Actually I did put them in their own respective uniforms before but I refactored it to the current version since the 3x3 matrix had some spare space that could be used for this purpose.

**Review by karimnaaji — APPROVED** [review]
👍 LGTM @akoylasar

**Pull request merged**
